text/0000-container-leak.md

```diff
@@ -0,0 +1,335 @@
+- Feature Name: `container-leak`
+- Start Date: 2020-08-01
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Describe a standard set of methods for converting container types like `Box<T>`, `Arc<T>`, `Vec<T>`, `String` to and from raw pointers.
+
+For containers with a single value like `Box<T>`, `Arc<T>`, and `Rc<T>`, any subset of the following method pairs should be added to work with their raw representations:
+
+- `leak`: leak the container and return an arbitrarily long-lived shared or mutable reference to its allocated content.
+- `leak_raw`: leak the container and return a `NonNull<T>` pointer to its content. The type `T` is the same as `Deref::Target`, so `Self::leak_raw(value)` is equivalent to `NonNull::from(&*self)` and `NonNull::from(Self::leak(value))`.
+- `unleak_raw`: take a previously leaked `NonNull<T>` pointer and restore the container from it.
+- `into_raw`: leak the container and return a raw pointer to its content.
+- `from_raw`: take a previously leaked raw pointer and restore the container from it.
+
+For multi-value containers like `Vec<T>` and `String`, any subset of the following method pairs should be added to work with their raw representations:
+
+- `leak`: shrink the container to its initialized length, leak it and return an arbitrarily long-lived shared or mutable reference to its allocated content.
```

Comment:
> shrink the container to its initialized length

Is shrinking it first the best option here? I'd personally expect `Vec::leak` to *not* shrink the contents. I think of `leak` as just something like *"hey this Vec you got here, never deallocate it"*, like `std::mem::forget`, except we can still access the leaked contents. Shrinking it first might involve a lot of copying, just to save some unused capacity. Without shrinking, it does what I'd expect from a `leak` function: it's basically a no-op other than stopping deallocation from happening.

If I want to not waste any unused capacity, I could still easily call `shrink_to_fit` first (or leak the `.into_boxed_slice()` instead).

If `leak` does shrink, I'd have to use unsafe code to leak it 'efficiently' (i.e. without copying/shrinking, but wasting some unused heap space).

Reply:
Do you have an example use-case where this inefficiency would matter? The best one I can think of is a startup latency sensitive embedded system, that might want to quickly leak a lot of data at startup. However that use-case might be better off using a separate "pointer-bump" allocator for data it wants to leak. Not possible yet. But with custom allocators it will be.

Other than that situation, I'd expect `shrink_to_fit` to almost always be the right choice, because the inefficiency of something done once at startup would be outweighed by the long-term gain of having a bit less memory pressure.

Reply:
It doesn't feel very much like Rust to hide a potentially complex operation (shrinking/reallocating) inside a function that appears to be cheap or even free (`leak()`). It might very well be the case that there are not many programs where it would matter much, but it'd be good if functions do what their name suggests. The name `leak` suggests it just leaks the memory allocation of the container. Not that it copies all the data over to another new allocation first and leaks that instead while deallocating the original allocation.

If `leak()` doesn't shrink, that's easily explained in the documentation, and shrink-leaking is still easy to achieve with by calling `shrink_to_fit()` first (or using `into_boxed_*().leak()`). (Or maybe a `shrink_and_leak()` can be added?)

If `leak()` doesn't just leak but also shrinks, then *just* leaking is hard to achieve. It'd require `unsafe {}` and a raw pointer.

I also think `leak()` should be consistent with `leak_raw`, which only leaks and doesn't shrink.

Reply:
> Is shrinking it first the best option here? I'd personally expect Vec::leak to not shrink the contents.

Maybe whether or not these methods drop extra capacity should be left unspecified? I think `Vec::leak` only shrinks as a consequence of being implemented through `Vec::into_boxed_slice` so that it can use `Box::leak`.

Reply:
>  I think of leak as just something like "hey this Vec you got here, never deallocate it"

Current leak implementations can return a lifetime shorter than `&'static`, which means at the end of its lifetime it would be legal to reconstitute the original allocation. For that one needs to know the capacity.

Reply:
> Current leak implementations can return a lifetime shorter than `&'static`, which means at the end of its lifetime it would be legal to reconstitute the original allocation.

As far as I know, the only reason it can return lifetimes shorter than `'static` is to make it possible to leak `Vec<Something<'notstatic>>`. Turning a leaked Vec back into the original Vec doesn't really seem like an intended use case for `leak()`. `into_raw_parts` seems like the right function in that case.

Reply:
Preferable the term leak should only be used in a context where the intent is to leak the vector, and it should not shrink the vector.

Different versions of `into_raw(_parts)`/`from_raw(_parts)` should generally be used to decompose this types if you later one intent to recompose them.

Except that sadly due to historic reason some of the `into_raw` methods do not return `NonNull` and as such require an otherwise unnecessary unsafe block (or `.unwrap()`)  to create a `NonNull`.

With regard to `shrink_to_fit()` we should warn with bold text in the leak documentation that reconstructing a `Vec` from which you only have the `len` and `data_ptr` will lead to unsafe behavior when dropping it as the `Layout` for `alloc` and `dealloc` do not match. That is why `Vec::into_boxed_slice()` does call `Vec::schrink_to_fit()` as else wise dropping the boxed slice would cause unsafe behavior of which the consequences are allocator defined. 

Using `leak()` to create a pointer the the underlying slice and then passing it to a C-FFI as `ptr`+`len` was a bug I just ran into...